content/program-security.md

```diff
@@ -0,0 +1,681 @@
+
+### How to create errors
+
+The `solana_program` crate includes a `ProgramError` enum with a list of generic errors that we can use, but it's often useful to create your own to provide more context about the error when debugging.
+
+We can define our own errors by creating an enum type that lists the errors we want to use. For example, the `NoteError` contains variants `Forbidden` and `InvalidLength`. The enum is made into a Rust `Error` type by using the `derive` attribute macro to implement the `Error` trait from the `thiserror` library. Each error type also has its own #[error(…)] error notation. This lets you provide an error message for each particular error type.
+
+```rust
+use solana_program::{program_error::ProgramError};
+use thiserror::Error;
+
+#[derive(Error)]
+pub enum NoteError {
+    #[error("Wrong note owner")]
+    Forbidden,
+
+    #[error("Text is too long")]
+    InvalidLength,
+}
+```
+
+### How to return errors
+
+The compiler expects errors returned by the program to be of type `ProgramError` from the `solana_program` crate. That means we won't be able to return our custom error unless we have a way to convert it into this type. The following implementation handles conversion between our custom error and the `ProgramError` type.
+
+```rust
+impl From<NoteError> for ProgramError {
+    fn from(e: NoteError) -> Self {
+        ProgramError::Custom(e as u32)
+    }
+}
+```
+
+To return the custom error from the program, simply use the `into()` method to convert the error into an instance of `ProgramError`.
+
+```rust
+if pda != *note_pda.key {
+    return Err(NoteError::Forbidden.into());
+}
+```
+
+## Basic security checks
+
+While these won't comprehensively secure your program, there are a few security checks you can keep in mind to fill in some of the larger gaps in your code:
+
+- Ownership checks - used to verify if an account is owned by the program
+- Signer checks - used to verify that account has signed a transaction
+- General Account Validation - used to verify an account is the expected account
+- Data Validation - used to verify the inputs provided by a user
+
+### Ownership checks
+
+An ownership check verifies that an account is owned by the expected public key. Let's use the note-taking app example that we've referenced in previous lessons. In this app, users can create, update, and delete notes that are stored by the program in PDA accounts.
+
+If a user were to invoke the `update` instruction, they would need supply the `pda_account` for the movie review they want to update. Since the user can input any instruction data they want, they could provide an account whose data matches the data format of a note account but was not created by the note-taking program. This account could potentially contain malicious data and so should not be trusted.
```

Comment:
This paragraph has all the right parts, but feels a little janky, ...needs to end on hoow the problem is introduced? 
 sugg:

> When a user invokes the `update` instruction, they also need to supply a `pda_account`.  We presume the provided `pda_account` is for the particular movie review they want to update, but the user can input any instruction data they want. They could even potentially send data which matches the data format of a note account but was not also created by the note-taking program.  This security vulnerability is one potential way to introduce malicious code.